File: backend/app/models/dashboard_config.py
from __future__ import annotations

from pydantic import BaseModel


class FilterOptionsSource(BaseModel):
    data_source_id: str
    value_column: str
    depends_on: dict[str, str] = {}


class FilterPresetOption(BaseModel):
    label: str
    value: int | str


class FilterConfig(BaseModel):
    id: str
    label: str
    type: str  # "single-select" | "multi-select" | "preset-range"
    lockable: bool = False
    options_source: FilterOptionsSource | None = None
    options: list[FilterPresetOption] | None = None
    default_value: int | str | None = None


class KpiSource(BaseModel):
    data_source_id: str
    metric: str


class KpiTrend(BaseModel):
    type: str  # "percentage_of"
    reference_kpi: str


class KpiConfig(BaseModel):
    id: str
    label: str
    format: str  # "number" | "currency" | "percent"
    sources: list[KpiSource]
    aggregation: str = "sum"
    trend: KpiTrend | None = None


class KpiSegment(BaseModel):
    kpi_id: str
    label: str
    color: str


class ChartLayout(BaseModel):
    col: int
    row: int
    width: int
    height: int


class ChartSource(BaseModel):
    data_source_id: str
    metric: str | None = None
    label: str | None = None


class DashboardChartConfig(BaseModel):
    id: str
    title: str
    type: str
    source_type: str = "query"  # "query" | "kpi_values"
    sources: list[ChartSource] | None = None
    kpi_segments: list[KpiSegment] | None = None
    layout: ChartLayout
    cross_filter: bool | None = None  # None = participates (default), False = opt-out
    drill_hierarchy: list[str] | None = None
    drill_detail_data_source_id: str | None = None


class GridSource(BaseModel):
    data_source_id: str


class GridColumn(BaseModel):
    field: str
    header: str
    type: str = "string"  # "string" | "number" | "date"


class VisibleWhen(BaseModel):
    kpi: str
    condition: str  # "gt" | "lt" | "eq"
    value: int | float


class GridConfig(BaseModel):
    id: str
    title: str
    data_source_id: str | None = None
    sources: list[GridSource] | None = None
    merge_on: list[str] | None = None
    merge_type: str | None = None
    columns: list[GridColumn]
    visible_when: VisibleWhen | None = None
    layout: ChartLayout
    cross_filter_column: str | None = None


class DashboardFeatures(BaseModel):
    cross_filter: bool = False
    drill_down: bool = False


class LayoutConfig(BaseModel):
    type: str = "flow"
    sections: list[str] = ["filters", "kpis", "charts", "grids"]


class DashboardConfig(BaseModel):
    id: str
    name: str
    description: str = ""
    features: DashboardFeatures = DashboardFeatures()
    filters: list[FilterConfig] = []
    kpis: list[KpiConfig] = []
    charts: list[DashboardChartConfig] = []
    grids: list[GridConfig] = []
    layout: LayoutConfig = LayoutConfig()

File: frontend/src/types/dashboard-config.ts
export interface FilterOptionsSource {
  dataSourceId: string
  valueColumn: string
  dependsOn: Record<string, string>
}

export interface FilterPresetOption {
  label: string
  value: number | string
}

export interface FilterConfig {
  id: string
  label: string
  type: 'single-select' | 'multi-select' | 'preset-range'
  lockable: boolean
  optionsSource?: FilterOptionsSource
  options?: FilterPresetOption[]
  defaultValue?: number | string
}

export interface KpiSource {
  dataSourceId: string
  metric: string
}

export interface KpiTrend {
  type: string
  referenceKpi: string
  /** Visual treatment for the trend badge.
   *  - 'delta' (default): time-series-style — coloured left border (green ↑ /
   *    red ↓), arrow icon, '+' prefix on positives. Use when the percentage
   *    represents a CHANGE (e.g. month-over-month).
   *  - 'ratio': neutral pill — muted background, no arrow, no '+' prefix, no
   *    border tint. Use when the percentage is a STATIC RATIO between two
   *    KPIs (e.g. break rate = breaks / records). Avoids implying a positive
   *    or negative movement when no movement is being expressed. */
  display?: 'delta' | 'ratio'
}

export interface KpiConfig {
  id: string
  label: string
  format: 'number' | 'currency' | 'percent'
  sources: KpiSource[]
  aggregation: string
  trend?: KpiTrend
  visibleWhen?: VisibleWhen
  /** Optional CSS custom-property name (e.g. `--chart-positive`, `--series-8`)
   * used as the card's accent — drives the left-border color, the trend pill
   * background tint, and the trend pill text color. When set it overrides the
   * default derivation (which colors the border from trend sign). Use this to
   * give domain meaning to a card regardless of its trend semantics — e.g.
   * "Breaks" stays amber even though its percentage_of value is positive. The
   * token must be defined in the theme's index.css. */
  accentColor?: string
}

export interface KpiSegment {
  kpiId: string
  label: string
  color: string
}

export interface ChartLayout {
  col: number
  row: number
  width: number
  height: number
}

export interface ChartSource {
  dataSourceId: string
  metric?: string
  label?: string
}

export interface DashboardChartConfig {
  id: string
  title: string
  type: string
  sourceType: 'query' | 'kpi_values'
  /** Reference to a chart in the chart library. When present, the view renderer
   *  fetches the stored chart config to resolve metricColumns, categoryColumn,
   *  and appearance — so the dashboard JSON doesn't need to duplicate them. */
  chartId?: string
  sources?: ChartSource[]
  kpiSegments?: KpiSegment[]
  layout: ChartLayout
  crossFilter?: boolean
  drillHierarchy?: string[]
  drillDetailDataSourceId?: string
  /** Per-chart refresh interval override in milliseconds. undefined = use dashboard default. */
  refreshInterval?: number
  /** Appearance overrides (colorRange, legend, labels) — passed through to chart renderer.
   *  Takes precedence over library chart appearance when both are present. */
  appearance?: {
    colorRange?: string[]
    showLegend?: boolean
    legendPosition?: 'top' | 'bottom' | 'left' | 'right'
    showXLabel?: boolean
    showYLabel?: boolean
    typeSpecific?: Record<string, unknown>
  }
  visibleWhen?: VisibleWhen
}

export interface GridColumn {
  field: string
  header: string
  type: 'string' | 'number' | 'date'
}

export interface GridSource {
  dataSourceId: string
}

export interface VisibleWhen {
  kpi: string
  condition: 'gt' | 'lt' | 'eq'
  value: number
}

export interface GridConfig {
  id: string
  title: string
  dataSourceId?: string
  sources?: GridSource[]
  mergeOn?: string[]
  mergeType?: string
  columns: GridColumn[]
  visibleWhen?: VisibleWhen
  layout: ChartLayout
  crossFilterColumn?: string
  /** When merging multiple sources, fill missing numeric cells in left-only /
   *  right-only rows with `0` instead of leaving them absent/null. Consumed
   *  by MergedSourceGrid → useDataSourceMerge → backend MergeRequest. */
  coalesceZero?: boolean
}

export interface DashboardFeatures {
  crossFilter: boolean
  drillDown: boolean
}

export interface DashboardLayoutConfig {
  type: string
  sections: string[]
}

export interface DashboardConfig {
  id: string
  name: string
  description: string
  features: DashboardFeatures
  filters: FilterConfig[]
  kpis: KpiConfig[]
  charts: DashboardChartConfig[]
  grids: GridConfig[]
  layout: DashboardLayoutConfig
  /** Auto-refresh interval in milliseconds. 0 = disabled. Default: 600000 (10 min). */
  autoRefreshInterval?: number
}

export interface KpiResult {
  id: string
  value: number
  percentage?: number
}

export interface KpisResponse {
  kpis: KpiResult[]
}

export interface DataSourceQueryResponse {
  columns: string[]
  rows: Record<string, unknown>[]
  rowCount: number
  truncated: boolean
}

export interface DistinctValuesResponse {
  values: string[]
}
